sawtooth-sdk-transaction-processor/src/main/java/sawtooth/sdk/processor/TransactionProcessor.java

```diff
@@ -14,252 +14,278 @@
-        builder.setStatus(TpProcessResponse.Status.INTERNAL_ERROR);
-        builder.setMessage(ie.getMessage());
-        if (ie.getExtendedData() != null) {
-          builder.setExtendedData(ByteString.copyFrom(ie.getExtendedData()));
-        }
-      }
-      stream.sendBack(
-          Message.MessageType.TP_PROCESS_RESPONSE,
-          message.getCorrelationId(),
-          builder.build().toByteString());
+  public final void setMaxOccupancy(final int max) {
+    this.maxOccupancy = max;
+  }
 
-    } catch (InvalidProtocolBufferException ipbe) {
-      LOGGER.info("Received Bytestring that wasn't requested that isn't TransactionProcessRequest");
+  /**
+   * Signal that this thread should stop.
+   */
+  public final void stopProcessor() {
+    try {
+      unregister();
+    } catch (InterruptedException exc) {
+      LOGGER.log(Level.INFO, "Interrupted while unregistering", exc);
+    } catch (TimeoutException exc) {
+      LOGGER.log(Level.INFO, "Timeout while unregistering", exc);
+    } catch (ValidatorConnectionError exc) {
+      LOGGER.log(Level.INFO, "Connection error while unregistering", exc);
     }
+    this.running.compareAndSet(true, false);
   }
 
   /**
    * Find the handler that should be used to process the given message.
-   *
-   * @param message The message that has the TpProcessRequest that the header that will be checked
-   *     against the handler.
+   * @param message The message that has the TpProcessRequest that the header that
+   *                will be checked against the handler.
    * @return the handler that should be used to processor the given message
    */
-  private TransactionHandler findHandler(final Message message) {
+  public final TransactionHandler findHandler(final Message message) {
     try {
-      TpProcessRequest transactionRequest =
-          TpProcessRequest.parseFrom(this.currentMessage.getContent());
+      TpProcessRequest transactionRequest = TpProcessRequest.parseFrom(message.getContent());
       TransactionHeader header = transactionRequest.getHeader();
-      for (int i = 0; i < this.handlers.size(); i++) {
-        TransactionHandler handler = this.handlers.get(i);
-        if (header.getFamilyName().equals(handler.transactionFamilyName())
-            && header.getFamilyVersion().equals(handler.getVersion())) {
-          return handler;
-        }
+      String familyName = header.getFamilyName();
+      String familyVersion = header.getFamilyVersion();
+      SimpleEntry<String, String> handlerKey = new SimpleEntry<>(familyName, familyVersion);
+      if (handlers.containsKey(handlerKey)) {
+        return handlers.get(handlerKey);
       }
       LOGGER.info("Missing handler for header: " + header.toString());
     } catch (InvalidProtocolBufferException ipbe) {
-      LOGGER.info("Received Message that isn't a TransactionProcessRequest");
-      ipbe.printStackTrace();
+      LOGGER.log(Level.INFO, "Received Message that isn't a TransactionProcessRequest", ipbe);
```

Comment:
Suggest: "Received message is not a TransactionProcessRequest"

Reply:
fixed